pkg/reconciler/topic/topic.go

```diff
@@ -0,0 +1,418 @@
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Topic %q became ready after %v", topic.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Topic", topic.Namespace, topic.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Topic, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) updateFinalizers(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, bool, error) {
+	source, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Don't modify the informers copy.
+	existing := source.DeepCopy()
+
+	var finalizers []string
+
+	// If there's nothing to update, just return.
+	exisitingFinalizers := sets.NewString(existing.Finalizers...)
+	desiredFinalizers := sets.NewString(desired.Finalizers...)
+
+	if desiredFinalizers.Has(finalizerName) {
+		if exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Add the finalizer.
+		finalizers = append(existing.Finalizers, finalizerName)
+	} else {
+		if !exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Remove the finalizer.
+		exisitingFinalizers.Delete(finalizerName)
+		finalizers = exisitingFinalizers.List()
+	}
+
+	mergePatch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers":      finalizers,
+			"resourceVersion": existing.ResourceVersion,
+		},
+	}
+
+	patch, err := json.Marshal(mergePatch)
+	if err != nil {
+		return desired, false, err
+	}
+
+	update, err := c.RunClientSet.PubsubV1alpha1().Topics(existing.Namespace).Patch(existing.Name, types.MergePatchType, patch)
+	return update, true, err
+}
+
+func addFinalizer(s *v1alpha1.Topic) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Insert(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func removeFinalizer(s *v1alpha1.Topic) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Delete(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *Reconciler) createOrUpdatePublisher(ctx context.Context, topic *v1alpha1.Topic) (*servingv1beta1.Service, error) {
+	existing, err := r.getPublisher(ctx, topic)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get an existing publisher", zap.Error(err))
+		return nil, err
+	}
+
+	desired := resources.MakePublisher(&resources.PublisherArgs{
+		Image:  r.publisherImage,
+		Topic:  topic,
+		Labels: resources.GetLabels(controllerAgentName, topic.Name),
+	})
+
+	if existing == nil {
+		svc, err := r.ServingClientSet.ServingV1beta1().Services(topic.Namespace).Create(desired)
+		logging.FromContext(ctx).Desugar().Info("Publisher created.", zap.Error(err), zap.Any("publisher", svc))
+		return svc, err
+	}
+
+	if diff := cmp.Diff(desired.Spec, existing.Spec); diff != "" {
```

Comment:
Use Semantic.DeepEqual here, not cmp.Diff. cmp.Diff is meant for tests. https://github.com/google/go-cmp

Reply:
https://github.com/GoogleCloudPlatform/cloud-run-events/issues/68